File: 2019/day07/day07-part2-a.py
from itertools import permutations

def parse_intcode(input_data):
    return list(map(int, input_data.split(',')))

def get_param(mode, param, intcode):
    if mode == 0:
        return intcode[param] if param < len(intcode) else 0
    elif mode == 1:
        return param

def ensure_memory_size(intcode, index):
    if index >= len(intcode):
        intcode.extend([0] * (index - len(intcode) + 1))
    return intcode

def run_intcode(intcode, inputs, pc=0):
    while pc < len(intcode):
        opcode = intcode[pc] % 100
        mode1 = (intcode[pc] // 100) % 10
        mode2 = (intcode[pc] // 1000) % 10

        param1 = intcode[pc + 1] if pc + 1 < len(intcode) else 0
        param2 = intcode[pc + 2] if pc + 2 < len(intcode) else 0
        param3 = intcode[pc + 3] if pc + 3 < len(intcode) else 0

        intcode = ensure_memory_size(intcode, param1)
        intcode = ensure_memory_size(intcode, param2)
        intcode = ensure_memory_size(intcode, param3)

        if opcode == 1:  # Addition
            intcode[param3] = get_param(mode1, param1, intcode) + get_param(mode2, param2, intcode)
            pc += 4
        elif opcode == 2:  # Multiplication
            intcode[param3] = get_param(mode1, param1, intcode) * get_param(mode2, param2, intcode)
            pc += 4
        elif opcode == 3:  # Input
            if inputs:
                intcode[param1] = inputs.pop(0)
            else:
                return ('await_input', intcode, pc)
            pc += 2
        elif opcode == 4:  # Output
            output = get_param(mode1, param1, intcode)
            return ('output', output, intcode, pc + 2)
        elif opcode == 5:  # Jump if true
            if get_param(mode1, param1, intcode) != 0:
                pc = get_param(mode2, param2, intcode)
            else:
                pc += 3
        elif opcode == 6:  # Jump if false
            if get_param(mode1, param1, intcode) == 0:
                pc = get_param(mode2, param2, intcode)
            else:
                pc += 3
        elif opcode == 7:  # Less than
            intcode[param3] = 1 if get_param(mode1, param1, intcode) < get_param(mode2, param2, intcode) else 0
            pc += 4
        elif opcode == 8:  # Equals
            intcode[param3] = 1 if get_param(mode1, param1, intcode) == get_param(mode2, param2, intcode) else 0
            pc += 4
        elif opcode == 99:  # Halt
            return ('halt', intcode)
        else:
            return ('error', f"Unknown opcode {opcode} at pc {pc}")

def amplifier_loop_sequence(intcode, phases):
    states = [(intcode[:], [phase], 0) for phase in phases]
    amp_index = 0
    signal = 0
    while True:
        state = states[amp_index]
        result = run_intcode(state[0], [signal] + state[1], state[2])

        if result[0] == 'output':
            signal = result[1]
            states[amp_index] = (result[2], [], result[3])
            amp_index = (amp_index + 1) % len(phases)
        elif result[0] == 'await_input':
            states[amp_index] = (result[1], [], result[2])
            amp_index = (amp_index + 1) % len(phases)
        elif result[0] == 'halt' and amp_index == 0:
            return signal
        else:
            return result

def find_best_phase_setting_with_feedback(intcode):
    phase_settings = permutations([5, 6, 7, 8, 9])
    max_signal = 0
    for phases in phase_settings:
        signal = amplifier_loop_sequence(intcode, phases)
        if isinstance(signal, int) and signal > max_signal:
            max_signal = signal
    return max_signal

# Read input from file
with open('day07-data.txt') as f:
    input_data = f.read().strip()

# Parse the input data
intcode = parse_intcode(input_data)

# Find the best phase setting with feedback loop
best_output = find_best_phase_setting_with_feedback(intcode)

# Print the result
print(f"Best output with feedback loop: {best_output}")

File: 2019/day07/day07-part2-a.exs
defmodule AdventOfCode2019.Day7 do
  def parse_intcode(input) do
    input
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
  end

  def get_param(mode, param, intcode) do
    case mode do
      0 -> Enum.at(intcode, param, 0)
      1 -> param
    end
  end

  def run_intcode(intcode, inputs, pc \\ 0, outputs \\ []) do
    opcode = rem(Enum.at(intcode, pc), 100)
    mode1 = rem(div(Enum.at(intcode, pc), 100), 10)
    mode2 = rem(div(Enum.at(intcode, pc), 1000), 10)
    
    param1 = Enum.at(intcode, pc + 1, 0)
    param2 = Enum.at(intcode, pc + 2, 0)
    param3 = Enum.at(intcode, pc + 3, 0)

    case opcode do
      1 ->
        intcode
        |> List.replace_at(param3, get_param(mode1, param1, intcode) + get_param(mode2, param2, intcode))
        |> run_intcode(inputs, pc + 4, outputs)
      
      2 ->
        intcode
        |> List.replace_at(param3, get_param(mode1, param1, intcode) * get_param(mode2, param2, intcode))
        |> run_intcode(inputs, pc + 4, outputs)
      
      3 ->
        [input | inputs_tail] = inputs
        intcode
        |> List.replace_at(param1, input)
        |> run_intcode(inputs_tail, pc + 2, outputs)
      
      4 ->
        output = get_param(mode1, param1, intcode)
        {:output, output, intcode, pc + 2, inputs, outputs}
      
      5 ->
        new_pc = if get_param(mode1, param1, intcode) != 0 do
          get_param(mode2, param2, intcode)
        else
          pc + 3
        end
        run_intcode(intcode, inputs, new_pc, outputs)
      
      6 ->
        new_pc = if get_param(mode1, param1, intcode) == 0 do
          get_param(mode2, param2, intcode)
        else
          pc + 3
        end
        run_intcode(intcode, inputs, new_pc, outputs)
      
      7 ->
        intcode
        |> List.replace_at(param3, if(get_param(mode1, param1, intcode) < get_param(mode2, param2, intcode), do: 1, else: 0))
        |> run_intcode(inputs, pc + 4, outputs)
      
      8 ->
        intcode
        |> List.replace_at(param3, if(get_param(mode1, param1, intcode) == get_param(mode2, param2, intcode), do: 1, else: 0))
        |> run_intcode(inputs, pc + 4, outputs)
      
      99 ->
        {:halt, Enum.reverse(outputs)}
      
      _ ->
        {:error, "Unknown opcode"}
    end
  end

  def permutations([]), do: [[]]
  def permutations(list) do
    for elem <- list, rest <- permutations(list -- [elem]), do: [elem | rest]
  end

  def amplifier_loop_sequence(intcode, phases) do
    # Initial amplifier states
    states = Enum.map(phases, fn phase ->
      {intcode, [phase], 0} # Initial state for each amplifier: intcode, input, and program counter (pc)
    end)

    # Feedback loop
    loop(states, 0, 0)
  end

  defp loop(states, current_amp, input) do
    {intcode, inputs, pc} = Enum.at(states, current_amp)
    {:output, output, new_intcode, new_pc, new_inputs, _} = run_intcode(intcode, [input | inputs], pc)

    # Update the state of the current amplifier
    new_states = List.replace_at(states, current_amp, {new_intcode, new_inputs, new_pc})

    # If we reached the last amplifier, the output feeds back into the first one
    next_amp = rem(current_amp + 1, length(states))

    # Check if the last amplifier has halted
    if current_amp == 4 and match?({:halt, _}, run_intcode(new_intcode, [], new_pc)) do
      output
    else
      loop(new_states, next_amp, output)
    end
  end

  def find_best_phase_setting_with_feedback(intcode) do
    [5, 6, 7, 8, 9]
    |> permutations()
    |> Enum.map(fn phases -> amplifier_loop_sequence(intcode, phases) end)
    |> Enum.max()
  end
end

# Read input from the file 'day07-data.txt'
input = File.read!("day07-data.txt")

# Parse the input data and find the best phase setting with feedback loop
intcode = AdventOfCode2019.Day7.parse_intcode(input)
best_output = AdventOfCode2019.Day7.find_best_phase_setting_with_feedback(intcode)

# Print the result
IO.puts("Best output with feedback loop: #{best_output}")
